lua/neo-tree/ui/selector.lua

```diff
@@ -0,0 +1,266 @@
+  local separator_config = utils.resolve_config_option(config, "source_selector", nil)
+  if separator_config == nil then
+    log.warn("Cannot find source_selector config. `create_selector` abort.")
+    return {}
+  end
+  local get_strlen = vim.api.nvim_strwidth
+  local text = separator_config.tab_labels[source_name] or source_name
+  local text_length = get_strlen(text)
+  if separator_config.tabs_min_width ~= nil and text_length < separator_config.tabs_min_width then
+    text = M.text_layout(text, separator_config.content_layout, separator_config.tabs_min_width)
+    text_length = separator_config.tabs_min_width
+  end
+  if separator_config.tabs_max_width ~= nil and text_length > separator_config.tabs_max_width then
+    text = M.text_layout(text, separator_config.content_layout, separator_config.tabs_max_width)
+    text_length = separator_config.tabs_max_width
+  end
+  local tab_hl = is_active and separator_config.highlight_tab_active
+    or separator_config.highlight_tab
+  local sep_hl = is_active and separator_config.highlight_separator_active
+    or separator_config.highlight_separator
+  return {
+    index = source_index,
+    is_active = is_active,
+    left = separator.left,
+    right = separator.right,
+    text = text,
+    tab_hl = tab_hl,
+    sep_hl = sep_hl,
+    length = text_length + get_strlen(separator.left) + get_strlen(separator.right),
+    text_length = text_length,
+  }
+end
+
+local text_with_hl = function(text, tab_hl)
+  if tab_hl == nil then
+    return text
+  end
+  return string.format("%%#%s#%s", tab_hl, text)
+end
+
+local add_padding = function(padding_legth, hl_padding, padchar)
+  if padchar == nil then
+    padchar = " "
+  end
+  return text_with_hl(string.rep(padchar, math.floor(padding_legth)), hl_padding)
+end
+
+M.text_layout = function(text, content_layout, output_width, text_length, hl_padding)
+  if text_length == nil then
+    text_length = vim.fn.strdisplaywidth(text)
+  end
+  local pad_length = output_width - text_length
+  local left_pad, right_pad = 0, 0
+  if pad_length < 0 then
+    return string.sub(text, 1, vim.str_byteindex(text, output_width)) -- lua string sub with multibyte seq
+  elseif content_layout == "start" then
+    left_pad, right_pad = 0, pad_length
+  elseif content_layout == "end" then
+    left_pad, right_pad = pad_length, 0
+  elseif content_layout == "center" then
+    left_pad, right_pad = pad_length / 2, math.ceil(pad_length / 2)
+  end
+  return add_padding(left_pad, hl_padding) .. text .. add_padding(right_pad, hl_padding)
+end
+
+M.render_tab = function(left_sep, right_sep, sep_hl, text, tab_hl, click_id)
+  local res = "%" .. click_id .. "@v:lua.___neotree_selector_click@"
+  if left_sep ~= nil then
+    res = res .. text_with_hl(left_sep, sep_hl)
+  end
+  res = res .. text_with_hl(text, tab_hl)
+  if right_sep ~= nil then
+    res = res .. text_with_hl(right_sep, sep_hl)
+  end
+  return res
+end
+
+M.create_selector = function(state, width)
```

Comment:
The components should just get the actual window width from vim.api.get_window_width instead of being passed this value, that way it can update dynamically as the window is resized. I can make this change in subsequent commits.

Reply:
I added several functions that calculates the window width with the api.
Commit: https://github.com/pysan3/neo-tree.nvim/commit/56dfe424b02eefb132fff266fb510c39d8671707

Please add the resizing features since it was too complicated for me to understand that part.

One thing I noticed that with `:set laststatus=3` (global statusline), the output might be a bit awkward since it loads with the width of the current window in mind.

- split
![image](https://user-images.githubusercontent.com/41065736/179665203-98a92347-144a-429c-84da-29abe686373d.png)
- floating window
![image](https://user-images.githubusercontent.com/41065736/179665389-8a94a24b-4faa-428c-857a-803dd1665c8d.png)

Reply:
Besides, I am not intending for `create_selector` to be used outside and just exported for people who might want to change the small details.
Instead I have added several wrapping functions that are meant to be used from outside. Here: 
https://github.com/pysan3/neo-tree.nvim/blob/56dfe424b02eefb132fff266fb510c39d8671707/lua/neo-tree/ui/selector.lua#L217
- `append_source_selector `
- `set_source_selector `
- `auto_set_source_selector `
- `return_source_selector `

Reply:
> Please add the resizing features since it was too complicated for me to understand that part.

You are almost there with `M.return_source_selector`, I'll just need to have it lookup it's own state so it can be called without arguments.

> One thing I noticed that with :set laststatus=3 (global statusline), the output might be a bit awkward since it loads with the width of the current window in mind.

The statusline option was not really meant for global statusline users. I figure each person is either a winbar/global statusline user and they would use the winbar option, or they are a statusline per window user and they will use the statusline option here.

Reply:
> I'll just need to have it lookup it's own state so it can be called without arguments.

I am not sure if we could do that. Currently `create_selector` relies on these elements, which I don't think is resolvable due to the following reasons.

- `state.name`: to see which tab is active, by comparing it with the source_name. For this one, I had no clue how to dodge it.
- `state.winid`: maybe possible to get current winid with lua api and so, but I didn't test it if it could function with `:Neotree action=show`. And also [this issue](https://github.com/nvim-neo-tree/neo-tree.nvim/pull/427#issuecomment-1187239881) ([solution](https://github.com/nvim-neo-tree/neo-tree.nvim/pull/427#issuecomment-1187255975)). At least, it didn't work with a simple `vim.api.nvim_win_call` since the rendering process hops around the current window and passed it to a fallback to nui to come back to the previous focused window? I think.

> I figure each person is either a winbar/global statusline user and they would use the winbar option, or they are a statusline per window user and they will use the statusline option here.

Yes, I think so too.

Reply:
> > I'll just need to have it lookup it's own state so it can be called without arguments.
> 
> I am not sure if we could do that

Don't worry, I know how to do that 😉 .

I'm going to merge this and denote it as experimental for now. I'll add in some of the changes I am looking for over the next few weeks before removing the "experimental" note and highlighting it on the README.

If anyone else wants to work on this, just let me know what you are doing so we don't duplicate the work.